crates/subspace-networking/src/create.rs

```diff
@@ -253,3 +421,60 @@ pub async fn create(
 
     Ok((node, node_runner))
 }
+
+// Builds the transport stack that LibP2P will communicate over along with an optional relay client.
+async fn build_transport(
+    keypair: identity::Keypair,
+    timeout: Duration,
+    yamux_config: YamuxConfig,
+    relay_transport: Option<ClientTransport>,
+) -> Result<Boxed<(PeerId, StreamMuxerBox)>, CreationError> {
+    let transport = {
+        let dns_tcp = TokioDnsConfig::system(TokioTcpConfig::new().nodelay(true))?;
+        let ws = WsConfig::new(TokioDnsConfig::system(TokioTcpConfig::new().nodelay(true))?);
+        let transport = dns_tcp.or_transport(ws);
+
+        // Add MemoryTransport to the chain to enable in-memory relay configurations.
+        MemoryTransport::default().or_transport(transport)
+    };
+
+    if let Some(relay_transport) = relay_transport {
+        let transport = OrTransport::new(relay_transport, transport);
+
+        Ok(upgrade_transport(
+            transport.boxed(),
+            &keypair,
+            timeout,
+            yamux_config,
+        ))
+    } else {
+        Ok(upgrade_transport(
+            transport.boxed(),
+            &keypair,
+            timeout,
+            yamux_config,
+        ))
+    }
```

Comment:
A bit cleaner:
```suggestion
    Ok(upgrade_transport(
        if let Some(relay_transport) = relay_transport {
            OrTransport::new(relay_transport, transport).boxed()
        } else {
            transport.boxed()
        },
        &keypair,
        timeout,
        yamux_config,
    ))
```

Or leave as is and make the whole `transport` argument generic in `upgrade_transport`, so you don't have to box it here at all (returned upgraded transport is boxed anyway and will result in the same type anyway, but one more pointer indirection will be eliminated).

Reply:
The proposed code is not compilable, I introduced complex generic arguments instead.